### ISSUE TITLE
Network Spec / Layers Documentation

### ISSUE THREAD

**Issue by chairbender** [tensorforce/tensorforce#307]
First of all, hello! I'm glad to have discovered this project and am planning on trying to use it.

As for my question - I am unable to find any documentation describing what each of the layers are, what they do, and what their parameters are. Have I missed it or is it nonexistent? If it doesn't exist, I'd be happy to add some.

**Comment by Mazecreator** [comment]
I was starting that but didn't get very far as I was trying to figure out the best way to present the information.  I have a few other things I am working on and more than willing to send you what I have already started (not much) if you want to take it on.  It goes in the docs directory to make sure it gets merged into the online manual.  Also, if you don't mind, the "summary_spec.md" title looks crazy in the online docs so if you put in a PR can you make the title shorter?  I was going to submit that with the network_specs.md.  I can do that separate if you would rather.

Also @krfricke helped me with the "summary_spec" formating as markdown doesn't work in the document system.  He sent me this information which seems to work well:
"I actually used pandoc for that, so pandoc summary_spec.md -o summary_spec.rst and then copy the tables from there into the .md file, enclosing them in the eval_rst code blocks"

[network_spec.zip](https://github.com/reinforceio/tensorforce/files/1648055/network_spec.zip)

**Comment by michaelschaarschmidt** [comment]
Just do add to this excellent starting point, the layers are really just standard neural network layers as you'd find them in  tensorflow, just custom implementations to deal with certain things the way we want (mostly managing shapes). Having them nicely documented would indeed be useful

**Comment by chairbender** [comment]
Thanks for the input. Yes, the TF layers are pretty standard. But there appears to be quite a few undocumented assumptions and other difficulties. For example, take the conv2d network. It fails if input is int32 - it requires float16 or float32. This was never a problem for me when using Keras - I think it took care of any needed conversions. Another difficulty is that it requires you to provide the inputs in BWHC format. Since my Gym environment obviously doesn't provide these extra dimensions for channel and batch in the observation, I needed to write a preprocessor to inject a dimension. I only found out about this stuff after digging into the source code when I got errors. Or maybe I am missing something.

As I use this more, I am noticing that the actual source code documentation for this project is actually quite good (especially for Python library standards). However, having to dig through the sources every time I want to use some functionality is not very sustainable. I wouldn't be able to use this library very effectively if I didn't have the excellent Pycharm IDE.

But, for some reason, despite the great source code documentation, it doesn't look like that is getting carried over to the online documentation on readthedocs. I've found the readthedocs site to be not very helpful.

So, IMO, maybe this is more of a problem of the readthedocs not having this useful information, as well as a few areas where source code is missing some documentation. For example, I see that every layer in layers.py has useful documentation describing the parameters, but that doesn't seem to show up in readthedocs. Other classes seem to have similar issues. The biggest pattern I see is a lack of showing the documentation for the instance variables and __init__ parameters.

**Comment by michaelschaarschmidt** [comment]
These are some good points, we will have to track down while some things are not getting displayed on readthedocs, which could also be a compilation issue. Just FYI, we are in the middle of a large update which will remove many structural inconsistencies (see branch memory here, it moves remaining pieces into TensorFlow), so it will take some time for us to get to documentation as we smooth out the issues before/after merging.

If you want to take charge on this particular issue (creating layer docs/improving inconsistencies), also feel free to join the dev chat for discussion specific issues. (https://docs.google.com/forms/d/1_UD5Pb5LaPVUviD0pO0fFcEnx_vwenvuc00jmP2rRIc/ via filling in form here). 

We will also likely provide integration with TensorFlow native layers following the branch merge, which has some advantages down the road. For example, some newer features in native TensorFlow such as the KFAC optimizer rely on specifics of native TensorFlow layers and distributions, and we don't want to lose out on these emerging features, so it's important to get this right.

**Comment by michaelschaarschmidt** [comment]
Also pinging @krfricke regarding readthedocs

**Comment by krfricke** [comment]
The source code documentation should be mirrored in the tensorforce package section of RTD: http://tensorforce.readthedocs.io/en/latest/tensorforce/tensorforce.html#subpackages

Currently, that's just an automatically generated documentation of the python docstrings (by using `sphinx-apidoc`) and not well integrated with the rest of the documentation. You're right that the init function was not always documented, I fixed that in d01719ecb6151458ed8ba1fbf6cfde068c3d4772, thanks for the suggestion. Also please note that the docs/tensorforce folder has to be updated any time a new subpackage is added (including some subpackage classes), so it might not be in sync all the time. I usually make sure to update that part of the documentation at least before a version release.

**Comment by AlexKuhnle** [comment]
@chairbender, just to let you know that I just (finally) added `tf.layers` support (available layers [here](https://github.com/reinforceio/tensorforce/blob/memory/tensorforce/core/networks/layer.py#L160)). For instance, the following specification of a dense layer is now possible: `dict(type='tf_layer', layer='dense', units=32)`.

**Comment by mysl** [comment]
@AlexKuhnle  may not be the best place to ask,  but it seems to me there are no examples/docs to demonstrate how to use the ComplexNetwork.  It might be a good idea to add an example for it ?
BTW,  thanks for open source this awesome project!

**Comment by AlexKuhnle** [comment]
@mysl, totally agree! This feature was actually contributed, and I have to admit, I so far haven't used it myself. :-)  Maybe @Mazecreator can provide an example here. But generally, yes, after sorting out the memory branch, documentation is high priority, since it seems the main pain point of using the framework.

**Comment by Mazecreator** [comment]
@mysl @chairbender was working on the specs so I didn't publish anything separately.  There are usually examples within the PR to describe how it works so you might be able to find more.  Here is an example with two state inputs mp & env and a few of the new activation features:

    network_spec_raw_test = [ 
        [
            dict(type='input', inputs=['mp','env']),
            dict(type='dense', size=None, activation={'name':'swish' ,'beta': 'learn'}, skip=True,l1_regularization=.00,l2_regularization=.000),
            dict(type='dense', size=15, activation={'name':'swish' ,'beta': 1.9}, skip=False,l1_regularization=.00,l2_regularization=.000)
        ],
        [
            dict(type='input', inputs=['mp','*']),                  
            dict(type='dense', size=None, activation='swish', skip=True,l1_regularization=.00,l2_regularization=.000),
            dict(type='dense', size=10, activation='swish', skip=False,l1_regularization=.00,l2_regularization=.000),
            dict(type='dense', size=10, activation='swish', skip=False,l1_regularization=.00,l2_regularization=.000),   
        ]    
    ]   

In this example, the first block concat's "mp & env" inputs into one input set.  It runs it through 2 dense layers.  The 2nd block then concat's "mp & (output of previous block)" and runs it through 3 dense layers.  So you can see "mp" is injected in two places in this net.

**Comment by mysl** [comment]
@AlexKuhnle  @Mazecreator ,  thanks for the reply and example!

**Comment by Mazecreator** [comment]
@mysl a few more examples can be found here as well:
https://github.com/reinforceio/tensorforce/pull/280

**Comment by jerabaul29** [comment]
What I will say may be stupid, but do you think there would be a way to interface tensorforce with Keras, so that one:

- 1 Builds an ANN in Keras (well documented, very nice API).
- 2 Imports it into a tensorforce agent.

Otherwise it feels like you will need to do quite a bit of work re-writing / copying the Keras wrappers around tensorflow, or am I wrong?

**Comment by mattdornfeld** [comment]
I think what @jerabaul29 brought up is a pretty important point. The fact that Tensorforce does not support building models through the Keras API is it's biggest weak point. Building networks through these custom JSON specifications drastically reduces reusability and extensibility of models built for the framework. Despite the fact that it's a much less mature project I'm using keras-rl right now, solely for the reason that I can specify my models using the Keras API. It would be great if Tensorforce could support this.

**Comment by AlexKuhnle** [comment]
I have never looked at it, but I guess it shouldn't be too problematic to integrate this. Tensorforce has its own JSON-based specification language for networks etc, but of course that only works within Tensorforce.

**Comment by Misterion777** [comment]
There is been a while since last comment, so I'll ask: still no way to import tensorflow/keras model to tensorforce?

**Comment by bezineb5** [comment]
I didn't try it myself, but there might be a way of doing that using SavableNetwork as described here: https://github.com/tensorforce/tensorforce/blob/master/FAQ.md#7-how-do-i-preload-values-in-my-model
It's probably easier said than done, though.

**Comment by AlexKuhnle** [comment]
Hey, thanks for the reminder. I've been mostly working on the revision version in the last months, but that would be a good feature to add. I haven't worked with Keras models, though, so if you have pointers and/or a few lines illustrating how you imagine it should work, that would be great.

**Comment by Misterion777** [comment]
@bezineb5 I guess even if we implement SavableNetwork, agents still need layers in that custom tensorforce json format, which doesn't support couple of things from TensorFlow (for example parameters for batch normalization or kernel initializers)
@AlexKuhnle just first thing that comes to mind is to change network_spec from json layers to some serialized object that can be obtained from saved tensorflow model

**Comment by AlexKuhnle** [comment]
Passing something else than a json is certainly possible. How would you expect to use it? Would it be something like a pretrained CNN feature extractor that is integrated as part of a policy network (so parts are still randomly initialized and trained from scratch), or would you import the entire policy network (including final mapping to actions layer?)? If the latter, how would you pretrain that "outside" of RL?

**Comment by Misterion777** [comment]
As I see it, it should be the first case. It is indeed not clear how we are supposed to pretrain it "outside" of RL.

**Comment by AlexKuhnle** [comment]
Improved
